File: PHP/Algorithms/ShellSort.php
<?php

class ShellSort {
    public static function sort(&$arr) {
        $arrLen = count($arr);
        $gap = 0;
        while($gap < (int)($arrLen / 3)) {
            $gap = $gap * 3 + 1;
        }

        while ($gap > 0) {
            for($i = $gap; $i < $arrLen; $i++) {
                $j = $i;
                $tmp = $arr[$i];
                while($j >= $gap && $arr[$j - $gap] > $tmp) {
                    $arr[$j] = $arr[$j - $gap];
                    $j = $j - $gap;
                }
                $arr[$j] = $tmp;
            }
            $gap = (int)(($gap - 1) / 3);
        }
        
    }
}

File: CSharp/Algorithms/ShellSort.cs
public static class ShellSort {
	
	public static void Sort(int[] arr) 
    {
		int gap = 0;
		while ( gap < arr.Length / 3 ) 
        {
			gap = gap * 3 + 1;
		}
		
		while ( gap > 0 ) {
			// Perform insertion sort on the gap.
			for ( int i = gap; i < arr.Length; i++ ) 
            {
				int j = i;
				int tmp = arr[i];
				// While there exists an inversion on the sub array..
				while ( j >= gap && arr[j - gap] > tmp )
                {
					// .. Shift the elements down..
					arr[j] = arr[j - gap];
					j = j - gap;
				}
				// .. And set the correct element.
				arr[j] = tmp;
				
			}
			gap = ( gap - 1 ) / 3;
		}
	}
}